File: src/poll_agents/email_service.py
"""Email verification service."""

import secrets
import string
from email.message import EmailMessage

import aiosmtplib

from .config.settings import SMTPSettings


class EmailService:
    """Service for sending verification emails."""

    def __init__(self, settings: SMTPSettings):
        self.settings = settings

    def generate_verification_code(self, length: int = 6) -> str:
        """Generate a random numeric verification code."""
        return "".join(secrets.choice(string.digits) for _ in range(length))

    async def send_verification_email(self, to_email: str, code: str) -> bool:
        """Send verification code via SMTP.

        Returns True if email sent successfully, False otherwise.
        """
        message = EmailMessage()
        message["From"] = self.settings.sender_email
        message["To"] = to_email
        message["Subject"] = "Poll Agents - Verification Code"
        message.set_content(f"""\
Welcome to Poll Agents!

Your verification code is: {code}

This code will expire in 5 minutes.

Thank you for participating.
""")

        try:
            await aiosmtplib.send(
                message,
                hostname=self.settings.host,
                port=self.settings.port,
                username=self.settings.username,
                password=self.settings.password,
                start_tls=self.settings.use_tls,
            )
            print(f"[EMAIL] Verification code sent to {to_email}")
            return True
        except Exception as e:
            print(f"[EMAIL ERROR] Failed to send to {to_email}: {e}")
            return False

File: src/email-service.ts
/** Email verification service. */

import nodemailer from "nodemailer";
import type { SMTPSettings } from "./config/settings";

export class EmailService {
  private transporter: nodemailer.Transporter;
  private senderEmail: string;

  constructor(settings: SMTPSettings) {
    this.senderEmail = settings.sender_email;
    this.transporter = nodemailer.createTransport({
      host: settings.host,
      port: settings.port,
      secure: settings.port === 465,
      auth: {
        user: settings.username,
        pass: settings.password,
      },
      connectionTimeout: 10000,
      greetingTimeout: 10000,
      socketTimeout: 10000,
      ...(settings.use_tls && settings.port !== 465 ? { requireTLS: true } : {}),
    });
  }

  generateVerificationCode(length: number = 6): string {
    const digits = "0123456789";
    let code = "";
    for (let i = 0; i < length; i++) {
      code += digits[crypto.getRandomValues(new Uint32Array(1))[0] % 10];
    }
    return code;
  }

  async sendVerificationEmail(toEmail: string, code: string): Promise<boolean> {
    const mailOptions = {
      from: this.senderEmail,
      to: toEmail,
      subject: "Poll Agents - Verification Code",
      text: `Welcome to Poll Agents!\n\nYour verification code is: ${code}\n\nThis code will expire in 5 minutes.\n\nThank you for participating.\n`,
    };

    try {
      await this.transporter.sendMail(mailOptions);
      console.log(`[EMAIL] Verification code sent to ${toEmail}`);
      return true;
    } catch (e) {
      console.log(`[EMAIL ERROR] Failed to send to ${toEmail}: ${e}`);
      return false;
    }
  }
}
